### PR TITLE
add a BreakpointMode for different types of breakpoints

### LINKED ISSUE #454: Add a "type" field for SourceBreakpoints
> Splitting off from https://github.com/microsoft/debug-adapter-protocol/issues/445

The original authors requested the capability to present hard vs soft breakpoints.

Hardware breakpoints are useful in a variety of low-level and embedded computing scenarios. They are supported on major architectures and by debugging such as gdb (the `hbreak` family of commands).

Unlike software breakpoints, they are usually limited in number. I do not think we need any new way to express this limit, as DA's can already return a `message` explaining why a Breakpoint may not have been verified. This I propose the following additions:

```ts
interface Capabilities {
  // ... in addition to exisitng properties:
  /**
   * Types of breakpoints supported by the debug adapter. If present, the client
   * may allow the user to select a type and include it in its `setBreakpoints` request.
   */
  breakpointTypes?: BreakpointType[];
  // ...
}

// borrowing properties from the `ExceptionBreakpointsFilter`:
interface BreakpointType {
  /**
   * The internal ID of the breakpoint type. This value is passed to the
   * `setBreakpoints` request.
   */
  type: string;

  /**
   * The name of the breakpoint type. This is shown in the UI.
   */
  label: string;

  /**
   * A help text providing additional information about the breakpoint type.
   * This string is typically shown as a hover and can be translated.
   */
  description?: string;
}

interface SourceBreakpoint {
  // ... in addition to exisitng properties:

  /**
   * The type of this breakpoint. This must be one of the `breakpointTypes` the
   * debug adapter advertised in its Capabilities.
   */
  type?: string;
}
```

### PR DESCRIPTION
**Pull request by connor4312**
Closes #454

### CODE REVIEW

File: debugAdapterProtocol.json [hunk 1]
```diff
@@ -3231,6 +3231,13 @@
 				"supportsSingleThreadExecutionRequests": {
 					"type": "boolean",
 					"description": "The debug adapter supports the `singleThread` property on the execution requests (`continue`, `next`, `stepIn`, `stepOut`, `reverseContinue`, `stepBack`)."
+				},
+				"breakpointModes": {
+					"type": "array",
+					"items": {
+						"$ref": "#/definitions/ChecksumAlgorithm"
```
Comment (kkistm) [inline debugAdapterProtocol.json:3238]: Should it be `#/definitiona/BreakpoingMode` instead?

File: specification.md [hunk 2]
```diff
@@ -3468,6 +3466,16 @@ interface Capabilities {
...
+
+  /**
+   * Modes of breakpoints supported by the debug adapter, such as 'hardware' or
+   * 'software'. If present, the client may allow the user to select a mode and
+   * include it in its `setBreakpoints` request.
+   * 
+   * Clients may present the first mode in this array as the 'default' mode in
+   * gestures that set breakpoints.
+   */
+  breakpointModes?: ChecksumAlgorithm[];
```
Comment (kkistm) [inline specification.md:3478]: Should it be `BreakpointMode[]` instead?

**Comment by gregg-miskelly** [comment]
Same suggestion as from the issue: I think we should add this to all of the breakpoint requests (missing: `FunctionBreakpoint`, `DataBreakpoint`).

**Comment by connor4312** [comment]
PR Updated.

**Comment by gregg-miskelly** [comment]
LGTM

File: specification.md [hunk 3]
```diff
@@ -4827,3 +4859,48 @@ export type InvalidatedAreas = 'all' | 'stacks' | 'threads' | 'variables'
...
+  mode: string;
+
+  /**
+   * The name of the breakpoint mode. This is shown in the UI.
+   */
+  label: string;
+
+  /**
+   * A help text providing additional information about the breakpoint mode.
+   * This string is typically shown as a hover and can be translated.
```
Comment (roblourens) [inline specification.md:4881]: 'label' could be translated too, right? Do we normally call this out?

File: debugAdapterProtocol.json [hunk 4]
```diff
@@ -4184,7 +4207,43 @@
...
+				"description": {
+					"type": "string",
+					"description": "A help text providing additional information about the breakpoint mode. This string is typically shown as a hover and can be translated."
+				},
+				"appliesTo": {
+					"type": "array",
+					"items": {
+						"$ref": "#/definitions/BreakpointModeApplicability"
+					},
+					"description": "Describes one or more type of breakpoint this mode applies to."
```
Comment (roblourens) [inline debugAdapterProtocol.json:4233]:
```suggestion
					"description": "Describes one or more types of breakpoints this mode applies to."
```

**Review by roblourens — APPROVED** [review]

**Pull request merged**